### DISCUSSION TITLE
Apply Singleton pattern to EuTxManager

### DISCUSSION THREAD

**Discussion by SeonJK** [euphony-io/euphony#259] [Ideas]
# AS-IS
While using EuTxManager with multiple instances, they cause wrong  execution.
For example, adding multiple EuTxManager due to add multiple EuPI, app cannot easily manage EuPI.

# TO-DO
By applying Singleton pattern, code prevents from several issues.

![image](https://user-images.githubusercontent.com/37062147/190893228-ba85b2bd-4f84-46e9-b720-9bca4ca1cbdc.png)

<br/>

I suggest singleton pattern by Bill Pugh solution. It makes sure Thread-safe.

Applying singleton pattern by @Volatile and @Synchronized causes performance decreasement.
However by Bill Pugh solution, EuTxManager class makes singleton through Holder class. Holder class makes sure to load once.

References:
- https://stackoverflow.com/questions/50771372/bill-pugh-singleton-why-thread-safe
- https://itzone.com.vn/en/article/bill-pugh-singleton-in-java-incredibly-simple/
- https://medium.com/swlh/singleton-design-pattern-with-kotlin-2e6c8d42fc11#:~:text=Singleton.INSTANCE.doSomeThin()-,we,-can%20create%20a

**Comment by yeonns** [comment]
@SeonJK Thank you for suggestion! Great diagram and description :+1: 
If you apply singleton to EuTxManager, it will be very helpful!
